docs/userguide/performance/testing.md

```diff
@@ -0,0 +1,97 @@
+
+This is important because KumoMTA works in a highly parallel fashion, with very granular queues created for the various combinations of campaign, tenant, destination domain, egress_source, and site_name required by your outgoing traffic, and if you send your tests from a single tenant or to a single destination you will not be able to tune your environment for production traffic because you will have a small handful of queues instead of thousands of queues and the server behavior will be very different.
+
+## Generating Traffic Using `traffic-gen`
+For cases where accurate simulation is not feasible, KumoMTA includes a "Traffic Generator" that can be use to send volume test mail for this purpose. The `traffic-gen` appends a known domain to all outbound mail that resolves to your own loopback address so that mail can be delivered, but will never deliver to real addresses:
+
+```bash
+sudo /opt/kumomta/sbin/traffic-gen --target <your.sink.server>:25 --concurrency 20000 --message-count 100000 --body-size 100000
+```
+
+For additional parameters for the `traffic-gen` utility see:
+```bash
+sudo /opt/kumomta/sbin/traffic-gen --help
+```
+
+The `traffic-gen` script is used internally by the KumoMTA team to test performance before each release.
+
+A more advanced example is as follows:
+
+```bash
+$ cat traffic-gen.sh
+#!/bin/bash
+
+exec traffic-gen \
+        --target localhost:2025 \
+        --keep-going \
+        --concurrency 400 \
+        --body-size 60kb \
+        --duration 1500 \
+        --domain-suffix 'testingdomain.tld' \
+        --domain aol.com:1.5 \
+        --domain bellsouth.net:0.2 \
+        --domain comcast.net:0.5 \
+        --domain gmail.com:67.4 \
+        --domain hotmail.com:8.8 \
+        --domain icloud.com:1.2 \
+        --domain live.com:0.3 \
+        --domain me.com:0.2 \
+        --domain msn.com:0.5 \
+        --domain orange.fr:0.4 \
+        --domain outlook.com:1.2 \
+        --domain sbcglobal.net:0.3 \
+        --domain yahoo.com:5.9 \
+        $@
+```
+
+The preceding example uses the domain argument to list the destination domains that should be generated and their relative weights.
+
+It is helpful to use [custom routing](https://docs.kumomta.com/userguide/policy/routing/) to configure the test server to route all messages to the sink server, with the sink configured to dev/null all messages. Modify the `init.lua` on the test server with the following:
+
+```bash
+kumo.on('smtp_server_message_received', function(msg)
+    msg:set_meta('queue', 'my.sink.server')
+end)
+```
+
+Alternatively you can use `iptables` to re-route traffic to a sink:
+
+```bash
+iptables -t nat -A OUTPUT -p tcp \! -d 192.168.1.0/24 --dport 25 -j DNAT --to-destination 127.0.0.1:2026
```

Comment:
```suggestion
```console
$ iptables -t nat -A OUTPUT -p tcp \! -d 192.168.1.0/24 --dport 25 -j DNAT --to-destination 127.0.0.1:2026
```